scipy/spatial/transform/_rotation.pyx

```diff
@@ -2806,54 +2886,257 @@ cdef class Rotation:
+            # The weighted average of these rotations will be the optimal
+            # rotation
+            rots = []
+            for ai, bi in zip(a_non_inf, b_non_inf):
+                # Calculate angles between the primary and secondary vectors
+                a_angle = atan2(_norm3(np.cross(a_inf, ai)), np.dot(a_inf, ai))
+                b_angle = atan2(_norm3(np.cross(b_inf, bi)), np.dot(b_inf, bi))
+
+                # If the primary and secondary vectors are not too close to
+                # have numerical stability issues, fit orthonormal frames and
+                # find the rotation between them
+                if a_angle >= atol and b_angle >= atol:
+                    A = np.array(_matrix_from_pri_sec_vectors(a_inf, ai))
+                    B = np.array(_matrix_from_pri_sec_vectors(b_inf, bi))
+                    R = cls.from_matrix(A.T @ B)
+
+                # If the primary and secondary vectors are too close, we
+                # consider them parallel. Pull out the primary vectors and
+                # find the shortest path rotation between them
+                else:
+                    cross = np.cross(b_inf, a_inf)  # has norm of sin(theta)
+                    theta = atan2(_norm3(cross), np.dot(a_inf, b_inf))
+                    if theta < 1e-3:  # small angle Taylor series approximation
+                        theta2 = theta * theta
+                        rot_vec = cross * (1 + theta2 / 6
+                                           + theta2 * theta2 * 7 / 360)
+                    else:
+                        rot_vec = cross * theta / sin(theta)
+                    R = cls.from_rotvec(rot_vec)
+                rots.append(R)
+
+            if N == 2:
+                # Shortcut for single secondary vector
+                R_opt = rots[0]
+            else:
+                # Do a weighted average of all the rotations. Note that in
+                # order to match the result from the Kabsch algorithm in the
+                # limit as one of the weights goes to infinity, we must also
+                # scale the rotations by the projection of the secondary
+                # vectors projected into the plane pierced by the primary
+                # vectors
+                Rs = cls.concatenate(rots)
+                was = np.linalg.norm(np.cross(a_non_inf, a_inf), axis=1)
+                wbs = np.linalg.norm(np.cross(b_non_inf, b_inf), axis=1)
+                ws = np.asarray(weights[~weight_is_inf] * was * wbs)
+                if all(ws == 0):
+                    raise ValueError("All non-infinite weights are zero")
+                R_opt = Rs.mean(weights=ws)
+
+            # Calculated the root sum squared distance. The error will be 0 for
+            # the infinite weight vector, so we force that to avoid issues when
+            # multiplying by infinity
+            weights_non_inf = weights.copy()
+            weights_non_inf[weight_is_inf] = 0
```

Comment:
There is an issue here if the corresponding a and b vectors are not of equal length. In that instance the error will not be 0, and this will not line up with the baseline algorithm in the limit. I don't have a good idea of how to avoid this, without changing the behavior everywhere to calculate rssd from the normalized vectors.